_posts/2015-04-17-Enums-match-mutation-and-moves.md

```diff
@@ -0,0 +1,706 @@
+
+First, let us carefully state the meaning of a *non-ref* binding:
+
+* When matching a value of type `T`, an identifier pattern `i` will, on
+  a successful match, *move* the value out of the original input and
+  into `i`. Thus we can always conclude in such a case that `i` has type
+  `T` (or more succinctly, "`i: T`").
+
+  For some types `T`, known as *copyable* `T` (also pronounced "`T`
+  implements `Copy`"), the value will in fact be copied into `i` for such
+  identifier patterns. (Note that in general, an arbitrary type `T` is not copyable.)
+
+  Either way, such pattern bindings do mean that the variable `i` has
+  *ownership* of a value of type `T`.
+
+Thus, the bindings of `payload` in `tree_weight_v2` both have type
+`i32`; the `i32` type implements `Copy`, so the weight is copied into
+`payload` in both arms.
+
+Now we are ready to state what a ref-binding is:
+
+* When matching an L-value of type `T`, a `ref`-pattern `ref i`
+  will, on a successful match, merely *borrow* a reference into the
+  matched data. In other words, a successful `ref i` match of a value of
+  type `T` will imply that `i` has the type of a *reference* to `T`
+  (or more succinctly, "`i: &T`").
+
+Thus, in the `Node` arm of
+`tree_weight_v2`, `left` will be a reference to the left-hand box (which
+holds a tree), and `right` will likewise reference the right-hand tree.
+
+We can pass these borrowed references to trees into the recursive calls to `tree_weight_v2`,
+as the code demonstrates.
+
+Likewise, a `ref mut`-pattern (`ref mut i`) will, on a successful
+match, borrow a *mutable reference* into the input: `i: &mut T`. This allows
+mutation and ensures there are no other active references to that data
+at the same time. A destructuring
+binding form like `match` allows one to take mutable references to
+disjoint parts of the data simultaneously.
+
+This code demonstrates this concept by incrementing all of the
+values in a given tree.
+
+```rust
+fn tree_grow(t: &mut BinaryTree) {
+    //          ^~~~~~~~~~~~~~~ `&mut`: we have exclusive access to the tree
+    match *t {
+        BinaryTree::Leaf(ref mut payload) => *payload += 1,
+        BinaryTree::Node(ref mut left, ref mut payload, ref mut right) => {
+            tree_grow(left);
+            *payload += 1;
+            tree_grow(right);
+        }
+    }
+}
+
+#[test]
+fn tree_demo_3() {
+    let mut tree = sample_tree();
+    tree_grow(&mut tree);
+    assert_eq!(tree_weight_v2(&tree), (2 + 3 + 4) + 5 + 6);
+}
+```
+
+Note that the code above now binds `payload` by a `ref mut`-pattern;
+if it did not use a `ref` pattern, then `payload` would be bound to a
+local copy of the integer, while we want to modify the actual integer
+*in the tree itself*. Thus we need a reference to that integer.
+
+Note also that the code is able to bind `left` and `right`
+simultaneously in the `Node` arm. The compiler knows that the two
+values cannot alias, and thus it allows both `&mut`-references to live
```

Comment:
One more use of "alias", I think this could be rephrased "the two references cannot refer to the same part/piece/place/value/storage" (having a hard time coming up with both a clear and correct term to use).